accepted/000070-cluster_awareness.md

```diff
@@ -0,0 +1,309 @@
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.remove_node <cluster name> <optional params>
+  ```
+
+* Node reboot: Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.reboot_node <cluster name> <optional params>
+  ```
+
+  The implementation of how a node is rebooted is left to the specific cluster provider manager.
+
+* Cluster upgrade: As every cluster has its peculiarities when upgrading the cluster nodes and Uyuni/SUSE Manager should be agnostic about the implementation strategy, Uyuni/SUSE Manager will invoke the cluster provider manager on the management node of the cluster and will pass any optional parameter (requested to the user) to the call using Salt grains or by using a Salt formula.
+
+  ```
+  # salt <mgmt node> <cluster provider mgr>.upgrade_cluster <cluster name> <optional params>
+  ```
+
+### Single cluster node management with Uyuni/SUSE Manager
+
+The following actions refer to the individual cluster node registered to Uyuni/SUSE Manager. The following actions must be implemented after the cluster overview and its corresponding actions are implemented.
+NOTE: registering an SES 6 node to Uyuni/SUSE Manager is not be possible as SES 6 management node is a Salt master. From SES 7 on (under development), Salt will be used for deployment only.
+
+#### Relaxing system lock
+
+In the cluster overview, when listing individual nodes of the cluster, Uyuni/SUSE Manager will also offer the possibility of registering an individual node to Uyuni/SUSE Manager (with System > Bootstrapping).
+
+It is common between all analyzed cluster products that rebooting and package modification/removal are two forbidden actions on all nodes of the cluster.
+
+By default, Uyuni/SUSE Manager will put the minion in system lock when one of the following product is detected upon registration:
+
+* CaaSP: /etc/products.d/caasp.prod + pattern `pattern-caasp-Node`
+* SES: /etc/products.d/ses.prod
+* SLE-HA/SHAP: /etc/products.d/SLES_SAP.prod, /etc/products.d/sle-ha.prod
+
+When a system is locked, no action will be executed. System lock may be disabled by the user by disabling the corresponding Salt Formula.
+
+It will be possible to define an allowed list of actions that Uyuni/SUSE Manager can issue on a specific node by implementing a combination of selective blackout (https://github.com/uyuni-project/uyuni-rfc/pull/31/) and package locking.
+
+A product can define a list of:
+- an allowed list of actions in Uyuni/SUSE Manager action terms
+- packages or patterns to lock at the minion level
+
+This information must be provided by the product and must be inserted into Uyuni/SUSE Manager database.
```

Comment:
For locked packages, yes, we can shift the request to the cluster provider manager (`salt <mgmt node> <cluster provider mgr>.locked_pages`).
We expect this list not to change so often and be identical for every cluster type: considering we also need to save Actions (see below) it is something we think goes hand-in-hand.

For the Actions (as in the Uyuni/SUSE Manager terms), it requires a deep understanding of how an Action is implemented at the Uyuni/SUSE Manager level and it is tightly coupled to it.
Conceptually, it makes sense to have it at the Uyuni/SUSE Manager level instead of the cluster provider manager. 
Moreover, we already have a concept of locked actions in the database (in the Management stack) that we should reuse for https://github.com/uyuni-project/uyuni-rfc/pull/31/ :

```
susemanager=# select * from rhnactiontype;
 id  |                      label                      |                                                name                                                | trigger_snapshot | unlocked_only 
-----+-------------------------------------------------+----------------------------------------------------------------------------------------------------+------------------+---------------
   1 | packages.refresh_list                           | Package List Refresh                                                                               | Y                | N
   2 | hardware.refresh_list                           | Hardware List Refresh                                                                              | N                | N
   3 | packages.update                                 | Package Install                                                                                    | Y                | Y
   4 | packages.remove                                 | Package Removal                                                                                    | Y                | Y
[...]
```

Reply:
I'll try to reformulate to see if I understand the matter right:

Would it be possible to store the list of packages to lock and the list of actions to lock, per product, in the respective cluster provider managers?

If so, would that be sufficient to make it so Java code is unaware of details about specific products?

Reply:
Regarding locked packages, yes, Java code will be unaware of the details.
Regarding the forbidden actions, those will be stored in the database.

Do you suggest to modify the RFC to reflect what we clarified here?

Reply:
> Regarding locked packages, yes, Java code will be unaware of the details.

Good.

> Regarding the forbidden actions, those will be stored in the database.

Reflection I leave to you: if such data about forbidden actions had its authoritative source in the cluster provider manager, it could be copied to the database via some mechanism and make Java code once again unaware of product-specific details.

I for one would be happy to see that happening, because it would mean we are one step closer to a "fully pluggable" mechanism where adding a cluster type, in principle, would not require anything but Salt/Python coding.

> Do you suggest to modify the RFC to reflect what we clarified here?

That can't hurt :wink:

Reply:
OK, I just rewrote a bit this section and I hope it clarifies the behavior in this area.